vignettes/namespaces.Rmd

```diff
@@ -0,0 +1,416 @@
+  xml2::read_xml() 
+xml2::xml_add_child(xml, "code_block", 
+  xmlns = "http://commonmark.org/xml/1.0", info = "{r}", "1 + rnorm(1)\n")
+xml
+xml2::xml_find_all(xml, "//md:code_block", ns)
+cat(xslt::xml_xslt(xml, sty))
+```
+
+It works, but let's take a look at our namespaces:
+
+```{r}
+xml2::xml_ns(xml)
+```
+
+Every node we add with an unnamed namespace adds another default and in the end, 
+if we are doing a lot of substitution, we can end up with hundreds of namespaces.
+
+
+## No Namespace?
+
+What if we just tried to use no namespace?
+
+```{r}
+xml <- commonmark::markdown_xml("this is a **test**") %>%
+  xml2::read_xml() %>%
+  xml2::xml_ns_strip()
+xml2::xml_add_child(xml, "code_block", info = "{r}", "1 + rnorm(1)\n")
+xml
+xml2::xml_find_all(xml, "//code_block")
+sty <- xml2::read_xml(tinkr::stylesheet())
+cat(xslt::xml_xslt(xml, sty))
+```
+
+We can now add new nodes and use XPath without namespace prefixes or objects,
+but we have lost the ability to use our stylesheet :(
+
+But! Maybe we can do this by adding the namespace at the last minute!
+
+```{r}
+xml2::xml_set_attr(xml, "xmlns", "http://commonmark.org/xml/1.0")
+xml
+cat(xslt::xml_xslt(xml, sty))
+```
+
+## Harnessing the power of namespaces
+
+When you know that namespaces with prefixes will only respond to nodes with that
+prefix and all other nodes have no namespace, then you can add in nodes that can
+serve as anchors in your document or hiding markdown elements. Let's say we
+wanted to hide all markdown elements _except_ for code blocks. One way we could
+do this is to set up a namespace and add a prefix to all non-code-block nodes:
+
+```{r}
+xml <- commonmark::markdown_xml("this is a **test**") %>%
+  xml2::read_xml() %>%
+  xml2::xml_ns_strip()
+xml2::xml_add_child(xml, "code_block", info = "{r}", "1 + rnorm(1)\n")
+xml
+# Set the prefixed namespace in your document
+xml2::xml_set_attr(xml, "xmlns:tnk", "https://docs.ropensci.org/tinkr")
+# Find all nodes that are not code blocks
+nocode <- xml2::xml_find_all(xml, ".//*[not(self::code_block)]")
+nocode
+# Change the namespace of these nodes
+purrr::walk(nocode, xml2::xml_set_namespace, "tnk", "https://docs.ropensci.org/tinkr")
+xml
+xml2::xml_set_attr(xml, "xmlns", "http://commonmark.org/xml/1.0")
+sty <- xml2::read_xml(tinkr::stylesheet())
+cat(xslt::xml_xslt(xml, sty))
+```
+
+## Conclusion
+
+While developing {tinkr} we[^1] struggled a lot with understanding namespaces.
+This guide was our attempt at demystifying working with namespaces in {xml2}.
+For the casual user of {tinkr} who is interested in extracting data from 
+markdown documents, this guide is not very useful, but we hope that this 
+guide provies useful for the user who wants to use this for cleaning and
+standardizing their markdown documents. 
+
+[reparse]: https://community.rstudio.com/t/adding-nodes-in-xml2-how-to-avoid-duplicate-default-namespaces/84870/2?u=zkamvar
+[^1]: Well, mostly just Zhian.
```

Comment:
The truth is _you_ took the time to understand them, I just had to read your vignette!